pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -70,126 +76,245 @@ var (
 	if preferred := pod.Spec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution; len(preferred) > 0 {
 		sort.Slice(preferred, func(i int, j int) bool { return preferred[i].Weight > preferred[j].Weight })
-		r = r.Add(preferred[0].Preference.MatchExpressions...)
+		requirements = append(requirements, preferred[0].Preference.MatchExpressions...)
 	}
 	// Select first requirement. An outer loop will iteratively remove OR requirements if unsatisfiable
 	if pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil &&
 		len(pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) > 0 {
-		r = r.Add(pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions...)
+		requirements = append(requirements, pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions...)
 	}
-	return r
+	return requirements
 }
 
-func (r Requirements) WellKnown() (requirements Requirements) {
-	for _, requirement := range r {
-		if WellKnownLabels.Has(requirement.Key) {
-			requirements = requirements.Add(requirement)
+// Compatible returns errors with detailed messages when requirements are not compatible
+// Compatible is non-commutative (i.e., A.Compatible(B) != B.Compatible(A))
+func (r *Requirements) Compatible(requirements *Requirements) (errs *apis.FieldError) {
+	if r == nil {
+		errs = errs.Also(apis.ErrInvalidValue("nil requirements", "values"))
+		return errs
+	}
+	allKeys := stringsets.NewString(r.Keys()...).Union(stringsets.NewString(requirements.Keys()...))
+	for key := range allKeys {
+		if err := r.isKeyDefined(key, requirements); err != nil {
+			errs = errs.Also(err)
+		}
+		if err := r.hasCommons(key, requirements); err != nil {
+			errs = errs.Also(err)
+		}
+		if err := r.satisfyExistOperators(key, requirements); err != nil {
+			errs = errs.Also(err)
 		}
 	}
-	return requirements
+	return errs
 }
 
-// Keys returns unique set of the label keys from the requirements
-func (r Requirements) Keys() []string {
-	keys := sets.NewString()
-	for _, requirement := range r {
-		keys.Insert(requirement.Key)
-	}
-	return keys.UnsortedList()
-}
-
-// Requirements for the provided key, nil if unconstrained
-func (r Requirements) Requirement(key string) sets.String {
-	var result sets.String
-	// OpIn
-	for _, requirement := range r {
-		if requirement.Key == key && requirement.Operator == v1.NodeSelectorOpIn {
-			if result == nil {
-				result = sets.NewString(requirement.Values...)
-			} else {
-				result = result.Intersection(sets.NewString(requirement.Values...))
+// isKeyDefined returns errors when the given requirements requires a key that is not defined by this requirements
+// This is a non-commutative function designed to catch the following cases:
+// * Pod Spec has a label selector but provisioner doesn't have the label
```

Comment:
This case is permitted right?  For example, if my provisioner does not specify a `karpenter.sh/capacity-type` requirement, then it is unconstrained, meaning that if a pod has a nodeSelector for `spot`, then we should successfully validate the Pod's requirements with the provisioner.

Reply:
I actually had the same confusion a while a go and finally it hit me that provisioner labels will be bypassed in this situation since a pod with a node-selector can be validated to any provisioner with or without a proper label value.

Logically, we want the provisioner requirements to be tight (i.e., if it is not defined, it is not allowed) but the pod requirements to be as broad as possible (e.g., if not defined, pod doesn't care). This is to make sure that the provisioners behave according to what is defined, but not to what is not defined. This also allow us to always pick the tightest fix provisioner in selection.